src/dune_engine/fs_memo.ml

```diff
@@ -268,12 +268,12 @@ let dir_exists path =
    instead. For now, we keep it here because it seems nice to group all tracked
    file system access functions in one place, and exposing an uncached version
    of [file_digest] seems error-prone. We may need to rethink this decision. *)
-let file_digest ?(force_update = false) path =
+let path_digest ?(force_update = false) path =
   if force_update then (
     Cached_digest.Untracked.invalidate_cached_timestamp path;
-    Fs_cache.evict Fs_cache.Untracked.file_digest path);
+    Fs_cache.evict Fs_cache.Untracked.path_digest path);
   let+ () = Watcher.watch ~try_to_watch_via_parent:true path in
```

Comment:
This is no longer correct, is it? We can't watch directories via parents. This is an example of the kind of errors we make when mixing files and directories in one type. I think we should split this into two functions to avoid such pitfalls.

Reply:
> This is no longer correct, is it? We can't watch directories via parents.

Could you remind me what's the problem with watching directories via parents again?

Reply:
It's simple really. If you are reading `dir_contents` of `some/dir`, you are not going to get an event for creation of file `some/dir/x` by watching `some`. You need to subscribe to `some/dir` directly.

Reply:
I ended up reverting my `file -> changes`. They weren't even needed, and probably made things more confusing. I still kept the `Fs_cache.Untracked.path_digest` because I think it better reflects what we're installing in that cache. For better or for worse, we currently store file/directory in one table, so I think it's best if the name would reflect that.